compiler/src/iree/compiler/Codegen/Common/GPU/GPUBubbleResourceCasts.cpp

```diff
@@ -0,0 +1,199 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Common/GPU/Passes.h"
+#include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUDialect.h"
+#include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUOps.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "iree/compiler/Dialect/TensorExt/IR/TensorExtOps.h"
+#include "iree/compiler/Dialect/TensorExt/IR/TensorExtTypes.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_GPUBUBBLERESOURCECASTSPASS
+#include "iree/compiler/Codegen/Common/GPU/Passes.h.inc"
+
+namespace {
+struct GPUBubbleResourceCastsPass final
+    : impl::GPUBubbleResourceCastsPassBase<GPUBubbleResourceCastsPass> {
+  void runOnOperation() override;
+};
+} // namespace
+
+/// Walk the producers of |input| and return the hal.interface.binding_subspan
+/// that produces it if the binding is readonly.
+static Operation *isTriviallyProducedByReadOnlyViewLike(Value input) {
+  Value next = input;
+  while (auto slice = next.getDefiningOp<tensor::ExtractSliceOp>()) {
```

Comment:
Are there other view-like ops this could be? reshape and friends?

Reply:
This is a big type switch on the ops I could think of for now, pad and reshape included down below. A notable omission is cast because that's better handled as a canonicalization due to the way we want static info to propagate across the IR.

A whitelist made more sense to me to start but if there is a sensible general alternative we can switch in the future.